File: sdk/typescript/packages/core/src/delegation.ts
/**
 * TOLAP Delegation Chain Validation (canonical spec §15.3)
 *
 * A chain records how authority reached the principal making a call: a human
 * delegates to an agent, which delegates to a sub-agent. The invariant is that each
 * hop may hold **less** authority than its parent and never more. Without it, a
 * sub-agent could declare any purpose it liked and the chain would be decoration.
 *
 * The chain is only worth validating because it is inside the signed bytes — see
 * `SecurityContext.delegationChain`. Validating an unsigned chain would be checking
 * the attacker's own arithmetic.
 *
 * This validates the chain's *internal consistency*. It cannot establish that the
 * first hop's purpose was honestly declared; that is the stated limitation in
 * spec §13.
 */

import { globToRegex } from "./resolution.js";
import type { AccessResult, DelegationHop } from "./types.js";

// ---------------------------------------------------------------------------
// Purpose-glob matching -- a THIRD dialect, deliberately
// ---------------------------------------------------------------------------
//
// `resolution.ts` documents at length why the enforcement dialect (`globMatch`,
// `*` -> `.*`, case-insensitive) and the source-pattern dialect
// (`sourcePatternMatch`, `*` -> `[^:]*`, case-insensitive) must not be unified. This
// is the same argument for a third case, and neither existing helper can be called
// as-is:
//
// - Both are **case-insensitive**. Purpose comparison is case-SENSITIVE everywhere
//   else in this feature (`purposeId` at resolution, `judge.model` at the gate), and
//   a case-insensitive test here would admit a child purpose `Campaign-X` under a
//   parent `campaign-x` while resolution refused the same string. One rule saying
//   yes and the other no about the same value is the divergence class the canonical
//   spec exists to prevent.
// - `sourcePatternMatch`'s `*` stops at `:`. Purpose identifiers are hyphen-
//   delimited (`campaign-x-overlap`), so a parent `campaign-*` must reach a child
//   `campaign-x-overlap`; `[^:]*` happens to do that today only because purposes
//   contain no colons, which is a coincidence rather than a rule.
//
// So the *dialect* is the enforcement one -- `*` crosses everything -- and only the
// case sensitivity differs. That single difference is expressed by recompiling
// `globToRegex`'s source without its `i` flag, rather than by hand-rolling a fourth
// translation, and the reason is ReDoS: a naive `*` -> `[\s\S]*` translation
// backtracks catastrophically (measured: `("*a" x 40) + "-x"` against 200 `a`s does
// not return inside a minute, and JavaScript's RegExp has no evaluation timeout to cut
// it short, which is where .NET's match timeout does the work instead). `globToRegex`
// already emits the atomic-group shape that makes each wildcard commit; borrowing its
// source inherits that guarantee instead of reproducing the bug beside it.

/**
 * Bound on the pattern and value lengths one purpose glob may span.
 *
 * A second ceiling on total work, on top of the atomic-group shape above: spec §13
 * records that Python and TypeScript bound pattern and input length where .NET applies
 * a regex timeout. Fail-closed, so an over-long value denies the hop.
 *
 * The schema caps a purpose identifier at 128 characters, so this is generous by an
 * order of magnitude and cannot refuse a legitimate value.
 */
const MAX_PURPOSE_LENGTH = 1024;

/**
 * Glob matching for purpose identifiers: `*` matches any run of characters, crossing
 * every separator, and the comparison is **case-sensitive**.
 *
 * `?` is inherited from the enforcement dialect as a single-character wildcard. That
 * is a consequence of reusing the hardened compiler rather than a designed feature,
 * and it is unobservable in practice: `security-context.schema.json` constrains a hop
 * purpose to `^[a-z0-9*][a-z0-9*-]*$`, so `?` cannot appear in a schema-valid one.
 *
 * There is deliberately no `try`/`catch` around the recompilation — `globToRegex`
 * escapes every metacharacter it does not expand, so its source is always a
 * well-formed pattern, and a catch for it would be unreachable. Unreachable defensive
 * code is worse than none: it reads as a handled case that no test can exercise.
 */
function purposeGlobMatch(pattern: string, value: string): boolean {
  if (pattern.length > MAX_PURPOSE_LENGTH || value.length > MAX_PURPOSE_LENGTH) {
    return false;
  }

  // `.source` and no flags: same pattern, case-sensitive. Reading the flag off and
  // filtering it would be the same thing said less directly, since `i` is the only
  // flag `globToRegex` sets.
  return new RegExp(globToRegex(pattern).source).test(value);
}

/**
 * Whether a child purpose sits within a parent's scope.
 *
 * Three ways in, and the third is the one that matters:
 *
 * 1. Exactly equal. Delegation without narrowing.
 * 2. The parent is a glob that matches the child, so `campaign-*` admits
 *    `campaign-x-overlap`. This is how a parent expresses "any purpose in this
 *    family".
 * 3. The child extends the parent on a `-` **segment boundary**, so `campaign-x`
 *    admits `campaign-x-overlap` but **not** `campaign-xyz-evil`.
 *
 * The third rule exists because a plain prefix test — the obvious implementation —
 * accepts `campaign-xyz-evil` under `campaign-x`. The two purposes are unrelated;
 * one merely starts with the other's characters. Requiring the boundary makes the
 * prefix mean what a reader assumes it means.
 */
function isWithinScope(childPurpose: string, parentPurpose: string): boolean {
  if (childPurpose === parentPurpose) return true;

  if (parentPurpose.includes("*")) {
    return purposeGlobMatch(parentPurpose, childPurpose);
  }

  return childPurpose.startsWith(`${parentPurpose}-`);
}

/**
 * Whether every element of `child` appears in `parent`, compared case-sensitively.
 *
 * A subset test rather than equality: a hop may hold fewer scopes than its parent.
 */
function isSubset(child: string[], parent: string[]): boolean {
  const available = new Set(parent);
  return child.every((scope) => available.has(scope));
}

// ---------------------------------------------------------------------------
// Public API
// ---------------------------------------------------------------------------

/**
 * Validate a delegation chain.
 *
 * @param chain
 * The hops, oldest first. `undefined`, empty, and single-hop chains are allowed:
 * there is no parent to widen against, so there is nothing to check. An absent
 * chain is not treated as suspicious because delegation is opt-in — every
 * pre-purpose context has none.
 *
 * @returns
 * An allow, or a denial naming the offending hop by index. The reason strings are
 * part of the contract; integrators log and branch on them.
 */
export function validateDelegationChain(
  chain: DelegationHop[] | undefined,
): AccessResult {
  if (chain === undefined || chain.length <= 1) return { allowed: true };

  for (let i = 0; i < chain.length - 1; i++) {
    const parent = chain[i];
    const child = chain[i + 1];

    // Either side absent adds no constraint. A hop that declares no purpose is not
    // claiming one, so there is nothing to narrow and nothing to exceed; the
    // purpose-scoped policy check at resolution is what refuses an undeclared
    // purpose, and doing it twice here would deny every legitimate partial chain.
    const parentPurpose = parent.declaredPurpose;
    const childPurpose = child.declaredPurpose;
    if (
      parentPurpose !== undefined &&
      parentPurpose !== "" &&
      childPurpose !== undefined &&
      childPurpose !== "" &&
      !isWithinScope(childPurpose, parentPurpose)
    ) {
      return {
        allowed: false,
        reason:
          `delegation hop ${i + 1} purpose '${childPurpose}' is not within ` +
          `parent scope '${parentPurpose}'`,
      };
    }

    // Subset, not equality: a hop may hold fewer scopes than its parent. An empty
    // parent set is therefore not "unrestricted" but "nothing left to pass on", so
    // any child scope exceeds it (spec §3). Absent on either side adds no
    // constraint, matching the purpose rule above -- so the test is against
    // `undefined`, never for emptiness.
    if (
      parent.scopeNarrowing !== undefined &&
      child.scopeNarrowing !== undefined &&
      !isSubset(child.scopeNarrowing, parent.scopeNarrowing)
    ) {
      return {
        allowed: false,
        reason: `delegation hop ${i + 1} scopes exceed parent delegation`,
      };
    }
  }

  return { allowed: true };
}

File: sdk/python/tolap-core/tolap_core/delegation.py
"""Delegation-chain narrowing (canonical-enforcement-spec.md section 15.3).

A chain records how authority reached the principal making a call: a human delegates
to an agent, which delegates to a sub-agent. The invariant is that each hop may hold
less authority than its parent and never more. Without it, a sub-agent could declare
any purpose it liked and the chain would be decoration.

The chain is only worth validating because it is inside the signed bytes -- see
:attr:`tolap_core.models.SecurityContext.delegation_chain`. Validating an unsigned
chain would check the attacker's own arithmetic.

This validates the chain's *internal consistency*. It cannot establish that the first
hop's purpose was honestly declared; that is the stated limitation in spec section 13.
"""

from __future__ import annotations

from tolap_core.enforcement import AccessResult
from tolap_core.models import DelegationHop


def validate_delegation_chain(chain: list[DelegationHop] | None) -> AccessResult:
    """Validate a delegation chain, oldest hop first.

    ``None``, empty, and single-hop chains are allowed: there is no parent to widen
    against, so there is nothing to check. An absent chain is not treated as
    suspicious because delegation is opt-in -- every pre-purpose context has none.

    Returns an allow, or a denial naming the offending hop by index. The reason
    strings are part of the contract; integrators log and branch on them.
    """
    if chain is None or len(chain) <= 1:
        return AccessResult(allowed=True)

    for index in range(len(chain) - 1):
        parent = chain[index]
        child = chain[index + 1]

        # Either side absent adds no constraint. A hop that declares no purpose is
        # not claiming one, so there is nothing to narrow and nothing to exceed; the
        # purpose-scoped policy check at resolution is what refuses an undeclared
        # purpose, and doing it twice here would deny every legitimate partial chain.
        # An empty string is treated as absent, matching how the signing projection
        # normalizes it.
        if (
            parent.declared_purpose
            and child.declared_purpose
            and not _is_within_scope(child.declared_purpose, parent.declared_purpose)
        ):
            return AccessResult(
                allowed=False,
                reason=(
                    f"delegation hop {index + 1} purpose '{child.declared_purpose}' "
                    f"is not within parent scope '{parent.declared_purpose}'"
                ),
            )

        # Subset, not equality: a hop may hold fewer scopes than its parent. An empty
        # parent set is therefore not "unrestricted" but "nothing left to pass on", so
        # any child scope exceeds it (spec section 3). Absent on either side adds no
        # constraint, matching the purpose rule above -- which is why this tests
        # ``is not None`` rather than truthiness.
        if (
            parent.scope_narrowing is not None
            and child.scope_narrowing is not None
            and not set(child.scope_narrowing) <= set(parent.scope_narrowing)
        ):
            return AccessResult(
                allowed=False,
                reason=f"delegation hop {index + 1} scopes exceed parent delegation",
            )

    return AccessResult(allowed=True)


def _is_within_scope(child_purpose: str, parent_purpose: str) -> bool:
    """Whether a child purpose sits within a parent's scope.

    Three ways in, and the third is the one that matters:

    1. Exactly equal. Delegation without narrowing.
    2. The parent is a glob that matches the child, so ``campaign-*`` admits
       ``campaign-x-overlap``. This is how a parent expresses "any purpose in this
       family".
    3. The child extends the parent on a ``-`` segment boundary, so ``campaign-x``
       admits ``campaign-x-overlap`` but **not** ``campaign-xyz-evil``.

    The third rule exists because a plain prefix test -- the obvious implementation,
    -- accepts
    ``campaign-xyz-evil`` under ``campaign-x``. The two purposes are unrelated; one
    merely starts with the other's characters. Requiring the boundary makes the prefix
    mean what a reader assumes it means.

    Case-sensitive, matching the ``purpose_id`` comparison at resolution and unlike
    both existing glob helpers in this SDK. Both choices deny a mis-cased purpose
    rather than admitting it, which is the direction that matters.
    """
    if child_purpose == parent_purpose:
        return True

    if "*" in parent_purpose:
        return _case_sensitive_glob_match(parent_purpose, child_purpose)

    return child_purpose.startswith(parent_purpose + "-")


def _case_sensitive_glob_match(pattern: str, value: str) -> bool:
    """Glob-match a purpose identifier: ``*`` matches any run of characters.

    Neither existing helper fits, which is why this is here rather than a call to
    one of them. :func:`tolap_core.resolution._compile_source_pattern` expands ``*``
    to ``[^:]*`` for colon-delimited source triples and is case-insensitive;
    :func:`tolap_core.enforcement._pattern_matches` expands it to ``.*`` but is also
    case-insensitive and treats brackets specially. Purpose identifiers are
    hyphen-delimited, not colon-delimited, and must compare case-sensitively -- so
    borrowing either would change the answer. The three dialects are deliberately
    separate rather than unified.

    Implemented by splitting on ``*`` and scanning forward rather than by compiling a
    regex. That is a **deliberate deviation from the .NET reference**, which uses a
    bounded ``Regex`` with a 100 ms match timeout because .NET has one and Python's
    ``re`` does not. A pattern such as ``*a*a*a...-x`` against a long run of ``a``
    backtracks catastrophically under ``re``, and there would be no timeout to catch
    it. This matcher cannot backtrack at all -- each ``*``-separated literal is
    located once, left to right -- so the pathological input returns a non-match in
    linear time instead of relying on a bound to fail closed. The observable
    behaviour is identical for every well-formed pattern, and strictly better for a
    hostile one.

    ``pattern`` must contain at least one ``*``; :func:`_is_within_scope` has already
    handled equality and only reaches here when one is present. There is deliberately
    no guard for the wildcard-free case: it would be unreachable, and unreachable
    defensive code is worse than none -- it reads as a handled case that no test can
    exercise.
    """
    parts = pattern.split("*")
    prefix, suffix = parts[0], parts[-1]
    if not value.startswith(prefix) or not value.endswith(suffix):
        return False

    start = len(prefix)
    end = len(value) - len(suffix)
    if start > end:
        # The anchors overlap, so no assignment of the wildcards can cover the value.
        return False

    for part in parts[1:-1]:
        found = value.find(part, start, end)
        if found < 0:
            return False
        start = found + len(part)

    return True
